File: packages/swift/Sources/VoicevoxCore/Voicevox.swift
import Foundation

/// 高レベル Facade。通常の利用はこのクラスだけで完結する。
///
/// ```swift
/// let voicevox = try await Voicevox()
///
/// // 1. モデル一覧
/// let models = try await voicevox.listModels()
///
/// // 2. 利用規約に同意してダウンロード(個別 or 一括)
/// voicevox.acceptLicense(modelId: "0")
/// try await voicevox.downloadModel(id: "0")
///
/// // 3. 合成
/// let wav = try await voicevox.synthesis(text: "こんにちは", modelId: "0")
/// ```
public actor Voicevox {
    private let synthesizer: Synthesizer
    private let manager: ModelManager
    private let catalog: LicenseCatalog
    /// モデルダウンロードの同時実行数上限。
    private let maxConcurrentDownloads: Int

    public let gate: LicenseGate

    /// - Parameters:
    ///   - openJtalkDictDir: Open JTalk 辞書ディレクトリ(nil なら同梱辞書)。
    ///   - onnxruntimePath: macOS 等での ONNX Runtime dylib パス(nil なら既定探索)。
    ///   - modelsDir: モデル保存先(nil なら Application Support)。
    ///   - maxConcurrentDownloads: 並列ダウンロードの同時実行数(既定4)。
    public init(
        openJtalkDictDir: URL? = nil,
        onnxruntimePath: String? = nil,
        modelsDir: URL? = nil,
        maxConcurrentDownloads: Int = 4
    ) throws {
        let catalog = try LicenseCatalog.bundled()
        let gate = LicenseGate(termsVersion: catalog.termsVersion)
        self.catalog = catalog
        self.gate = gate
        self.manager = try ModelManager(catalog: catalog, gate: gate, modelsDir: modelsDir)
        self.synthesizer = try Synthesizer(
            openJtalkDictDir: openJtalkDictDir,
            onnxruntimePath: onnxruntimePath
        )
        self.maxConcurrentDownloads = max(1, maxConcurrentDownloads)
    }

    // MARK: - モデル一覧

    /// 利用可能な全モデルの一覧(ダウンロード状態付き)。
    public func listModels() -> [VoicevoxModelInfo] {
        catalog.models.map { model in
            var m = model
            m.isDownloaded = manager.isDownloaded(modelId: model.id)
            return m
        }
    }

    /// 全モデル共通の利用規約(VOICEVOX 音声モデル利用規約)のURL。
    public nonisolated var termsURL: String { catalog.termsURL }

    // MARK: - ライセンス同意

    /// モデルの利用規約に同意する。アプリ側は規約を提示した上で呼ぶこと。
    public nonisolated func acceptLicense(modelId: String) {
        gate.accept(modelId: modelId)
    }

    /// 同意状態を確認する。
    public nonisolated func isLicenseAccepted(modelId: String) -> Bool {
        gate.isAccepted(modelId: modelId)
    }

    // MARK: - ダウンロード

    /// 1モデルをダウンロードする(要同意)。ダウンロード済みなら何もしない。
    ///
    /// - Parameter onProgress: 進捗コールバック。(受信済みbytes, 全体bytes)。
    ///   高頻度で呼ばれるので、UI 更新は自前でスロットルすること。
    public func downloadModel(
        id: String,
        onProgress: (@Sendable (Int64, Int64) -> Void)? = nil
    ) async throws {
        try await manager.download(modelId: id, onProgress: onProgress)
    }

    // MARK: - 削除

    /// スタイルID を含むモデルのID。
    ///
    /// スタイルID だけを保持しているアプリから `synthesis` の modelId を解決するのに使う。
    /// 一覧表示だけなら `VoicevoxCatalog` を使えばネイティブ初期化を待たずに済む。
    public nonisolated func modelId(forStyle styleId: UInt32) -> String? {
        catalog.models.first { model in
            model.characters.contains { $0.styles.contains { $0.id == styleId } }
        }?.id
    }

    /// ダウンロード済みモデルのローカルサイズ(bytes)。未ダウンロードなら 0。
    public func downloadedSize(modelId: String) -> Int64 {
        guard let url = try? manager.localURL(for: modelId),
              let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
        else { return 0 }
        return (attributes[.size] as? NSNumber)?.int64Value ?? 0
    }

    /// ダウンロード済みモデルの合計サイズ(bytes)。
    public func downloadedSize() -> Int64 {
        catalog.models.reduce(0) { $0 + downloadedSize(modelId: $1.id) }
    }

    /// ダウンロード済みモデルを削除する。ロード済みならアンロードしてから消す。
    /// 未ダウンロードなら何もしない。
    public func deleteModel(_ modelId: String) throws {
        let info = try manager.info(for: modelId)
        if synthesizer.isLoaded(modelId: modelId) {
            try synthesizer.unloadVoiceModel(modelId: modelId, vvmId: info.vvmId)
        }
        try manager.remove(modelId: modelId)
    }

    /// 複数モデルを並列ダウンロードする。
    /// - Returns: モデルIDごとの成否。一部失敗でも throw せず結果で返す。
    @discardableResult
    public func downloadModels(ids: [String]) async -> [String: Result<Void, VoicevoxError>] {
        let semaphore = AsyncSemaphore(value: maxConcurrentDownloads)
        let manager = self.manager
        return await withTaskGroup(
            of: (String, Result<Void, VoicevoxError>).self,
            returning: [String: Result<Void, VoicevoxError>].self
        ) { group in
            for id in ids {
                group.addTask {
                    await semaphore.wait()
                    defer { Task { await semaphore.signal() } }
                    do {
                        try await manager.download(modelId: id)
                        return (id, .success(()))
                    } catch let error as VoicevoxError {
                        return (id, .failure(error))
                    } catch {
                        return (id, .failure(.downloadFailed(modelId: id, underlying: String(describing: error))))
                    }
                }
            }
            var results: [String: Result<Void, VoicevoxError>] = [:]
            for await (id, result) in group {
                results[id] = result
            }
            return results
        }
    }

    /// 全モデルを一括ダウンロードする(全モデルへの同意が必要)。
    @discardableResult
    public func downloadAllModels() async -> [String: Result<Void, VoicevoxError>] {
        await downloadModels(ids: catalog.models.map(\.id))
    }

    // MARK: - 合成

    /// テキストから音声(WAV)を合成する。
    ///
    /// モデルが未ダウンロードの場合は `modelNotDownloaded` を投げる
    /// (暗黙のダウンロードは行わない)。未ロードならロードしてから合成する。
    /// - Parameters:
    ///   - text: 読み上げるテキスト。
    ///   - modelId: モデルID(`listModels()` の `id`)。
    ///   - styleId: スタイルID(`characters[].styles[]` から選ぶ)。
    ///     nil なら最初の talk スタイル。talk スタイルを持たないモデル
    ///     (歌唱合成用の s0 など)は `talkNotSupported` を投げる。
    public func synthesis(text: String, modelId: String, styleId: UInt32? = nil) async throws -> Data {
        let info = try manager.info(for: modelId)
        guard info.supportsTalk else {
            throw VoicevoxError.talkNotSupported(modelId: modelId)
        }
        guard manager.isDownloaded(modelId: modelId) else {
            throw VoicevoxError.modelNotDownloaded(modelId: modelId)
        }
        try gate.require(modelId: modelId)

        if !synthesizer.isLoaded(modelId: modelId) {
            try synthesizer.loadVoiceModel(at: manager.localURL(for: modelId), modelId: modelId)
        }

        guard let style = styleId
            ?? info.characters.lazy.flatMap(\.talkStyles).first?.id
        else {
            throw VoicevoxError.talkNotSupported(modelId: modelId)
        }
        let synthesizer = self.synthesizer
        // 合成はCPU負荷が高いのでactorの外(バックグラウンド)で実行する
        return try await Task.detached(priority: .userInitiated) {
            try synthesizer.tts(text: text, styleId: style)
        }.value
    }
}

File: packages/android/lib/src/main/kotlin/jp/voicevox/android/Voicevox.kt
package jp.voicevox.android

import android.content.Context
import java.io.File
import jp.hiroshiba.voicevoxcore.blocking.Onnxruntime
import jp.hiroshiba.voicevoxcore.blocking.OpenJtalk
import jp.hiroshiba.voicevoxcore.blocking.Synthesizer
import jp.hiroshiba.voicevoxcore.blocking.VoiceModelFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

/**
 * 高レベル Facade。通常の利用はこのクラスだけで完結する。
 *
 * ```kotlin
 * val voicevox = Voicevox.create(context)
 *
 * // 1. モデル一覧
 * val models = voicevox.listModels()
 *
 * // 2. 利用規約に同意してダウンロード(個別 or 一括)
 * voicevox.acceptLicense("0")
 * voicevox.downloadModel("0")
 *
 * // 3. 合成
 * val wav = voicevox.synthesis("こんにちは", modelId = "0")
 * ```
 */
class Voicevox private constructor(
    private val catalog: LicenseCatalog,
    val gate: LicenseGate,
    private val manager: ModelManager,
    private val synthesizer: Synthesizer,
    private val maxConcurrentDownloads: Int,
) {
    private val loadedModelIds = mutableSetOf<String>()
    private val loadMutex = Mutex()

    companion object {
        /**
         * Facade を構築する。ネイティブ初期化と辞書展開を含むため suspend。
         *
         * @param context アプリケーションコンテキスト。
         * @param modelsDir モデル保存先(null なら filesDir/voicevox/models)。
         * @param maxConcurrentDownloads 並列ダウンロードの同時実行数(既定4)。
         */
        suspend fun create(
            context: Context,
            modelsDir: File? = null,
            maxConcurrentDownloads: Int = 4,
        ): Voicevox = withContext(Dispatchers.IO) {
            val appContext = context.applicationContext
            val catalog = Json.decodeFromString<LicenseCatalog>(
                appContext.assets.open("licenses.json").bufferedReader().readText()
            )
            val gate = LicenseGate(appContext, catalog.termsVersion)
            val manager = ModelManager(
                catalog = catalog,
                gate = gate,
                modelsDir = modelsDir ?: File(appContext.filesDir, "voicevox/models"),
            )

            val dictDir = extractOpenJtalkDict(appContext)
            val onnxruntime = Onnxruntime.loadOnce().perform()
            val synthesizer = Synthesizer.builder(onnxruntime, OpenJtalk(dictDir.absolutePath)).build()

            Voicevox(catalog, gate, manager, synthesizer, maxConcurrentDownloads.coerceAtLeast(1))
        }

        /** assets の Open JTalk 辞書を filesDir に展開する(展開済みならスキップ)。 */
        private fun extractOpenJtalkDict(context: Context): File {
            val dictDir = File(context.filesDir, "voicevox/open_jtalk_dic")
            val marker = File(dictDir, ".complete")
            if (marker.exists()) return dictDir

            dictDir.mkdirs()
            val assets = context.assets
            for (name in assets.list("open_jtalk_dic").orEmpty()) {
                assets.open("open_jtalk_dic/$name").use { input ->
                    File(dictDir, name).outputStream().use { output -> input.copyTo(output) }
                }
            }
            marker.createNewFile()
            return dictDir
        }
    }

    // --- モデル一覧 ---

    /** 利用可能な全モデルの一覧(ダウンロード状態付き)。 */
    fun listModels(): List<VoicevoxModelInfo> =
        catalog.models.map { it.copy(isDownloaded = manager.isDownloaded(it.id)) }

    /** 全モデル共通の利用規約(VOICEVOX 音声モデル利用規約)のURL。 */
    val termsURL: String get() = catalog.termsURL

    // --- ライセンス同意 ---

    /** モデルの利用規約に同意する。アプリ側は規約を提示した上で呼ぶこと。 */
    fun acceptLicense(modelId: String) = gate.accept(modelId)

    /** 同意状態を確認する。 */
    fun isLicenseAccepted(modelId: String): Boolean = gate.isAccepted(modelId)

    // --- ダウンロード ---

    /** 1モデルをダウンロードする(要同意)。ダウンロード済みなら何もしない。 */
    suspend fun downloadModel(id: String) = manager.download(id)

    /**
     * 複数モデルを並列ダウンロードする(同時実行数は maxConcurrentDownloads)。
     * @return モデルIDごとの成否。一部失敗でも throw しない。
     */
    suspend fun downloadModels(ids: List<String>): Map<String, Result<Unit>> = coroutineScope {
        val semaphore = Semaphore(maxConcurrentDownloads)
        ids.map { id ->
            async {
                id to semaphore.withPermit { runCatching { manager.download(id) } }
            }
        }.awaitAll().toMap()
    }

    /** 全モデルを一括ダウンロードする(全モデルへの同意が必要)。 */
    suspend fun downloadAllModels(): Map<String, Result<Unit>> =
        downloadModels(catalog.models.map { it.id })

    // --- 合成 ---

    /**
     * テキストから音声(WAV)を合成する。
     *
     * モデルが未ダウンロードの場合は [VoicevoxException.ModelNotDownloaded] を投げる
     * (暗黙のダウンロードは行わない)。未ロードならロードしてから合成する。
     *
     * @param text 読み上げるテキスト。
     * @param modelId モデルID([listModels] の id)。
     * @param styleId スタイルID(characters[].styles[] から選ぶ)。null なら最初の talk スタイル。
     *   talk スタイルを持たないモデル(歌唱合成用の s0 など)は
     *   [VoicevoxException.TalkNotSupported] を投げる。
     */
    suspend fun synthesis(text: String, modelId: String, styleId: Int? = null): ByteArray {
        val info = manager.info(modelId)
        if (!info.supportsTalk) {
            throw VoicevoxException.TalkNotSupported(modelId)
        }
        if (!manager.isDownloaded(modelId)) {
            throw VoicevoxException.ModelNotDownloaded(modelId)
        }
        gate.require(modelId)

        loadMutex.withLock {
            if (modelId !in loadedModelIds) {
                withContext(Dispatchers.IO) {
                    VoiceModelFile(manager.localFile(modelId).absolutePath).use { model ->
                        synthesizer.loadVoiceModel(model)
                    }
                }
                loadedModelIds.add(modelId)
            }
        }

        val style = styleId
            ?: info.characters.asSequence()
                .flatMap { it.talkStyles }
                .firstOrNull()?.id
            ?: throw VoicevoxException.TalkNotSupported(modelId)
        // 合成はCPU負荷が高いのでDefaultディスパッチャで実行
        return withContext(Dispatchers.Default) {
            synthesizer.tts(text, style).perform()
        }
    }
}
